add alibi position embedding and support baichuan

qyccc opened the pull request:
This adds the [ALiBi](https://arxiv.org/abs/2108.12409) method and its flash attention version (using triton) for positional information. And it supports baichuan model trainig by porting over the implementation from [baichuan-inc/Baichuan2-13B-Base](https://huggingface.co/baichuan-inc/Baichuan2-13B-Base/blob/main/config.json) .

qyccc committed:
Update transformer.py

add alibi position embedding

qyccc committed:
Create pretrain_baichuan.py

qyccc committed:
Create baichuan_model.py

qyccc committed:
Create baichuan_checkpoint_conversion.py

qyccc committed:
Create baichuan_hf_to_megatron.sh

qyccc committed:
Update baichuan_hf_to_megatron.sh

qyccc committed:
Create baichuan_megatron_to_hf.sh

qyccc committed:
Update README.md

qyccc committed:
Update README.md

qyccc committed:
Update README_zh.md

qyccc committed:
Update README_zh.md

qyccc committed:
Update README.md

qyccc committed:
Update arguments.py

qyccc committed:
add collate_fn

qyccc committed:
Update training.py

qyccc committed:
add collate_fn

qyccc committed:
add alibi

qyccc committed:
Update tokenizer.py

qyccc committed:
Update transformer.py

qyccc committed:
Create Baichuan_13_standalone.sh

qyccc committed:
Update README.md

qyccc committed:
Update __init__.py

qyccc committed:
solve Unsupported gpu architecture 'compute_90'

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/alibaba/Megatron-LLaMA?pullRequest=54) <br/>All committers have signed the CLA.

li-yi-dong commented:
Cool! It may take some time to review 🙃

li-yi-dong commented on examples/Baichuan_13_standalone.sh:
```diff
@@ -0,0 +1,90 @@
+GLOBAL_BATCH_SIZE=$((($WORLD_SIZE * $MICRO_BATCH_SIZE) / ($TP_SIZE * $PP_SIZE) * 8))
+# GLOBAL_BATCH_SIZE=128
+
+JOB_NAME="LLaMA_tp${TP_SIZE}_pp${PP_SIZE}_mbs${MICRO_BATCH_SIZE}_gpus${WORLD_SIZE}"
```
Baichuan

li-yi-dong commented on megatron/arguments.py:
```diff
@@ -526,6 +526,9 @@ def _add_network_size_args(parser):
                        'attention. This is set to '
                        '   args.hidden_size // args.num_attention_heads '
                        'if not provided.')
+    group.add_argument('--position-embedding-type', type=str, default='learned_absolute',
```
May put together with `--use-rotary-position-embeddings`

li-yi-dong commented on megatron/fused_kernels/__init__.py:
```diff
@@ -23,9 +23,9 @@ def load(args):
     if int(bare_metal_major) >= 11:
         cc_flag.append('-gencode')
         cc_flag.append('arch=compute_80,code=sm_80')
-        if int(bare_metal_minor) >= 7:
```
I guess the minor version here is inaccurate. But simply delete the branch and never build for sm_90 is not good.
You may refer to this one https://github.com/NVIDIA/Megatron-LM/commit/a8affeae6e0932f96c5a2f5c8478123ca292aa85, or, I can merge it later.

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -16,17 +16,30 @@
+
 try:
     from flash_attn.flash_attn_interface import flash_attn_unpadded_func
+    from flash_attn.flash_attn_triton import flash_attn_func
```
I think it's better to separate Flash-Attn V1 /  Flash-Attn V2 and flash_attn_triton here.

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -43,6 +56,48 @@
         hyperparameters: transformer hyperparameters
 """
 
+def get_slopes(n):
```
You may put this in a separate file.

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -43,6 +56,48 @@
+
+
+def _gen_alibi_mask(num_attention_heads, max_seq_len):
+    slopes = torch.Tensor(get_slopes(num_attention_heads))
```
You may initialize these Tensors on GPU to avoid unnecessary HtoD and CPU/GPU sync.
This may help a lot for long sequence.

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -253,7 +313,7 @@ def forward(self, query_layer, key_layer,
         # ===================================
         # Raw attention scores. [b, np, s, s]
         # ===================================
-
```
Keep the blink line.

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -267,12 +327,14 @@ def forward(self, query_layer, key_layer,
+        # preallocting input tensor: [b * np, sq, sk]
         # preallocting input tensor: [b * np, sq, sk]
         matmul_input_buffer = mpu.get_global_memory_buffer().get_tensor(
-            (output_size[0]*output_size[1], output_size[2], output_size[3]),
```
Indent.

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -368,8 +434,8 @@ def forward(self, q, k, v):
         seqlen_k = k.shape[1]
+        # q, k, v = [rearrange(x, 'b s ... -> (b s) ...') for x in [q, k, v]]
 
-        q, k, v = [rearrange(x, 'b s ... -> (b s) ...') for x in [q, k, v]]
```
Flash-Attn V1 would be broken.

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -471,10 +542,24 @@ def __init__(self, init_method,
         self.checkpoint_core_attention = args.recompute_granularity == 'selective'
+        
+        self.apply_query_key_layer_scaling = args.apply_query_key_layer_scaling
+        world_size = mpu.get_tensor_model_parallel_world_size()
```
tensor_parallel_size

qyccc replied:
> tensor_parallel_size

sorry, I didn't get it. Do you mean the variance name should be tensor_parallel_size？

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -471,10 +542,24 @@ def __init__(self, init_method,
         if self.use_flash_attn:
             self.core_attention_flash = FlashSelfAttention(
-                causal=True, attention_dropout=args.attention_dropout
+                causal=True, softmax_scale=(1.0/self.norm_factor), attention_dropout=args.attention_dropout
```
This would broke LLaMA in a tricky way.

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -497,7 +582,7 @@ def custom_forward(*inputs):
             attention_mask = inputs[3]
             output_ = self.core_attention(query_layer, key_layer,
-                                          value_layer, attention_mask)
+                                            value_layer, attention_mask)
```
Why?

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -857,7 +927,7 @@ def forward(self, hidden_states, attention_mask,
         # MLP.
         mlp_output, mlp_bias = self.mlp(layernorm_output)
-
+        #     torch.distributed.barrier()
```
Delete

li-yi-dong commented on megatron/tokenizer/tokenizer.py:
```diff
@@ -551,7 +551,7 @@ def __init__(self, tokenizer_name_or_path, vocab_extra_ids=0):
         if vocab_extra_ids > 0:
             hf_tokenizer_kwargs["additional_special_tokens"] = [f"<extra_id_{_id}>" for _id in range(vocab_extra_ids)]
-        self.tokenizer = AutoTokenizer.from_pretrained(tokenizer_name_or_path, **hf_tokenizer_kwargs)
+        self.tokenizer = AutoTokenizer.from_pretrained(tokenizer_name_or_path, trust_remote_code=True,**hf_tokenizer_kwargs)
```
It should not default to True.

li-yi-dong reviewed (changes requested):
I'll review again once you solve the comments.

qyccc committed:
Update Baichuan_13_standalone.sh

qyccc committed:
Update arguments.py

qyccc committed:
Update transformer.py

qyccc committed:
Update utils.py

qyccc committed:
Update tokenizer.py

qyccc committed:
Update tokenizer.py

qyccc committed:
Update arguments.py

qyccc committed:
Update Baichuan_13_standalone.sh

qyccc committed:
Update training.py

qyccc committed:
support flash attention v1

qyccc committed:
Update Baichuan_13_standalone.sh

qyccc committed:
Update arguments.py

qyccc committed:
initialize certain Tensors on GPU

qyccc commented:
@li-yi-dong  Thanks for your time and cautious review. I have made the necessary changes and addressed the comments you mentioned. Please take another look at the updated version at your convenience.

qyccc committed:
Update transformer.py

qyccc committed:
Update __init__.py

qyccc committed:
Update Baichuan_13_standalone.sh

li-yi-dong commented on megatron/training.py:
```diff
@@ -56,6 +56,7 @@ def pretrain(train_valid_test_dataset_provider,
              forward_step_func,
              process_non_loss_data_func=None,
              extra_args_provider=None,
+             collator=None, 
```
Never been used. The newest Megatron-LM also removed.

li-yi-dong commented on megatron/training.py:
```diff
@@ -956,11 +962,11 @@ def build_train_valid_test_data_loaders(
+    # torch.distributed.broadcast(flags, 
+    #                             mpu.get_tensor_model_parallel_src_rank(), 
+    #                             group=mpu.get_tensor_model_parallel_group())
+    torch.distributed.broadcast(flags, 0)
```
You may start another PR for cherry-pick this. Meanwhile, you should cherry-pick the whole commit.

li-yi-dong commented on pretrain_baichuan.py:
```diff
@@ -0,0 +1,122 @@
+
+
+if __name__ == "__main__":
+    os.environ['CUDA_DEVICE_MAX_CONNECTIONS'] = '1'
```
Setting this ENV in python script here will not change the behavior of GPU driver.

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -523,7 +594,6 @@ def forward(self, hidden_states, attention_mask,
                 encoder_output=None, inference_params=None,
                 rotary_pos_emb=None):
         # hidden_states: [sq, b, h]
-
```
Keep the blink lines 🥺

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -857,7 +919,7 @@ def forward(self, hidden_states, attention_mask,
         # MLP.
         mlp_output, mlp_bias = self.mlp(layernorm_output)
-
+        
```
🥺

li-yi-dong commented on megatron/model/transformer.py:
```diff
@@ -1222,11 +1286,106 @@ def set_input_tensor(self, input_tensor):
         forward_step_func"""
         self.input_tensor = input_tensor
 
+    def _build_alibi_tensor(self, tensor, max_seq_len, num_attention_heads):
```
Placing this func together with `alibi_mask_func`

qyccc replied:
This func requires the internal variable first_run, so it cannot be placed in the utils.

li-yi-dong reviewed (changes requested):
Big thanks to your efforts and patience.
I added some comments to resolve.

qyccc committed:
remove collate_fn

qyccc committed:
Update indentent in transformer.py

qyccc committed:
Update data_samplers.py

qyccc committed:
Update training.py

qyccc committed:
Update transformer.py

qyccc committed:
Update transformer.py

qyccc committed:
Update utils.py

qyccc committed:
Update transformer.py

qyccc committed:
Update transformer.py
